### ISSUE TITLE
PGO upgrade from 4.6.0 to 4.7.2 via helm fails

### ISSUE THREAD

**Issue by keithlayne** [CrunchyData/postgres-operator#2650]
## Overview

Upgrading from 4.6.0 to 4.7.2 fails in post-upgrade script.

## Environment

Please provide the following details:

- Platform: `Kubernetes`
- Platform Version: `1.20.7`
- PGO Image Tag: `centos8-13.3-4.7.2`

### REPRO

1. Pull the helm chart:
    ```sh
    curl -L https://github.com/CrunchyData/postgres-operator/archive/v4.7.2.tar.gz \
      | tar zxv --strip-components 2 --wildcards "*/installers/helm"
    ```  
1. Attempt helm upgrade (re-using my previous values):
    ```sh
    mv helm helm-4.7.2
    helm upgrade --install postgres-operator helm-4.7.2 --values values.yaml --namespace pgo --atomic
    ```  

### EXPECTED

Operator upgrades cleanly.

### ACTUAL

```
Error: UPGRADE FAILED: release postgres-operator failed, and has been rolled back 
due to atomic being set: post-upgrade hooks failed: job failed: BackoffLimitExceeded
```

From the logs below, there are two `kubectl` commands that fail:

1. `kubectl delete clusterrolebinding pgo-cluster-admin`
    error: `Error from server (NotFound): clusterrolebindings.rbac.authorization.k8s.io "pgo-cluster-admin" not found`
    This seems to be ignored, so I don't think it's a problem. The name of the clusterrolebinding is not configurable via the helm chart.
1. `kubectl label secret generic -n pgo pgo-backrest-repo-config vendor=crunchydata`
    error: `Error from server (NotFound): secrets "generic" not found"`
    This looks like a mistake. AFAICT pgo doesn't create a secret called "generic". This looks like a typo from copying `kubectl create secret generic ...`.
    
At this point, my pgo installation is completely broken. Due to the `--atomic` flag on helm, it rolls back to the previous release, but I ended up without the operator deployment recreated. Fortunately, this was recoverable by running `helm upgrade` with the 4.6.0 version of the chart.

## Logs

<details>
 <summary>pgo-deploy-upgrade logs</summary>

```
nss_wrapper: user exists
nss_wrapper: group exists
nss_wrapper: environment configured

PLAY [Deploy Crunchy PostgreSQL Operator] **************************************

TASK [Gathering Facts] *********************************************************
ok: [localhost]

TASK [pgo-preflight : include_tasks] *******************************************
included: /ansible/postgres-operator/roles/pgo-preflight/tasks/vars.yml for localhost

TASK [pgo-preflight : Check for "/conf/values.yaml"] ***************************
ok: [localhost]

TASK [pgo-preflight : fail] ****************************************************
skipping: [localhost]

TASK [pgo-preflight : include_vars] ********************************************
ok: [localhost]

TASK [pgo-preflight : fail] ****************************************************
skipping: [localhost]

TASK [pgo-preflight : Check Operating System] **********************************
changed: [localhost]

TASK [pgo-preflight : assert] **************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [pgo-preflight : assert] **************************************************
skipping: [localhost]

TASK [pgo-preflight : assert] **************************************************
skipping: [localhost]

TASK [pgo-preflight : assert] **************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [pgo-preflight : include_tasks] *******************************************
skipping: [localhost]

TASK [pgo-preflight : include_tasks] *******************************************
included: /ansible/postgres-operator/roles/pgo-preflight/tasks/check_kubernetes.yml for localhost

TASK [pgo-preflight : Check if the kubectl command is installed] ***************
changed: [localhost]

TASK [pgo-preflight : Ensure kubectl is installed] *****************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [pgo-preflight : include_tasks] *******************************************
included: /ansible/postgres-operator/roles/pgo-preflight/tasks/check_vars.yml for localhost

TASK [pgo-preflight : Check if mandatory variables are defined] ****************
skipping: [localhost] => (item=pgo_operator_namespace) 
skipping: [localhost] => (item=pgo_installation_name) 
skipping: [localhost] => (item=pgo_admin_username) 
skipping: [localhost] => (item=pgo_admin_role_name) 
skipping: [localhost] => (item=pgo_admin_perms) 
skipping: [localhost] => (item=ccp_image_prefix) 
skipping: [localhost] => (item=ccp_image_tag) 
skipping: [localhost] => (item=pgo_image_prefix) 
skipping: [localhost] => (item=pgo_image_tag) 
skipping: [localhost] => (item=disable_auto_failover) 
skipping: [localhost] => (item=badger) 
skipping: [localhost] => (item=metrics) 
skipping: [localhost] => (item=archive_mode) 
skipping: [localhost] => (item=archive_timeout) 
skipping: [localhost] => (item=db_password_length) 
skipping: [localhost] => (item=create_rbac) 
skipping: [localhost] => (item=db_port) 
skipping: [localhost] => (item=db_replicas) 
skipping: [localhost] => (item=db_user) 
skipping: [localhost] => (item=backrest_storage) 
skipping: [localhost] => (item=backup_storage) 
skipping: [localhost] => (item=primary_storage) 
skipping: [localhost] => (item=replica_storage) 
skipping: [localhost] => (item=pgo_client_version) 
skipping: [localhost] => (item=pgbadgerport) 
skipping: [localhost] => (item=exporterport) 
skipping: [localhost] => (item=scheduler_timeout) 
skipping: [localhost] => (item=namespace_mode) 
skipping: [localhost] => (item=reconcile_rbac) 

TASK [pgo-operator : Set output directory fact] ********************************
ok: [localhost]

TASK [pgo-operator : Ensure output directory exists] ***************************
changed: [localhost]

TASK [pgo-operator : include_tasks] ********************************************
skipping: [localhost] => (item=openshift_auth.yml) 
skipping: [localhost] => (item=openshift.yml) 

TASK [pgo-operator : include_tasks] ********************************************
included: /ansible/postgres-operator/roles/pgo-operator/tasks/kubernetes_auth.yml for localhost
included: /ansible/postgres-operator/roles/pgo-operator/tasks/kubernetes.yml for localhost

TASK [pgo-operator : Set the Kubernetes Context] *******************************
skipping: [localhost]

TASK [pgo-operator : Get Namespace Details] ************************************
changed: [localhost]

TASK [pgo-operator : Create PGO Namespace] *************************************
skipping: [localhost]

TASK [pgo-operator : include_tasks] ********************************************
included: /ansible/postgres-operator/roles/pgo-operator/tasks/cleanup.yml for localhost

TASK [pgo-operator : Use kubectl or oc] ****************************************
ok: [localhost]

TASK [pgo-operator : Get number of Operator instances in this cluster] *********
changed: [localhost]

TASK [pgo-operator : Set boolean for multi-operator environment] ***************
ok: [localhost]

TASK [pgo-operator : Find watched namespaces] **********************************
changed: [localhost]

TASK [pgo-operator : Set Watched Namespaces] ***********************************
ok: [localhost]

TASK [pgo-operator : Delete Operator Deployment] *******************************
changed: [localhost]

TASK [pgo-operator : Delete existing ConfigMaps] *******************************
changed: [localhost]

TASK [pgo-operator : Delete existing secrets (Operator Namespace)] *************
changed: [localhost]

TASK [pgo-operator : Delete existing Services] *********************************
changed: [localhost]

TASK [pgo-operator : Delete existing Service Account (Operator Namespace)] *****
changed: [localhost]

TASK [pgo-operator : Delete existing Cluster Role Bindings] ********************
changed: [localhost] => (item=pgo-cluster-role)

TASK [pgo-operator : Delete cluster-admin Cluster Role Binding for PGO Service Account] ***
fatal: [localhost]: FAILED! => {"changed": true, "cmd": ["kubectl", "delete", "clusterrolebinding", "pgo-cluster-admin"], "delta": "0:00:00.085119", "end": "2021-08-28 16:31:27.626087", "msg": "non-zero return code", "rc": 1, "start": "2021-08-28 16:31:27.540968", "stderr": "Error from server (NotFound): clusterrolebindings.rbac.authorization.k8s.io \"pgo-cluster-admin\" not found", "stderr_lines": ["Error from server (NotFound): clusterrolebindings.rbac.authorization.k8s.io \"pgo-cluster-admin\" not found"], "stdout": "", "stdout_lines": []}
...ignoring

TASK [pgo-operator : Delete existing Cluster Roles] ****************************
changed: [localhost] => (item=pgo-cluster-role)

TASK [pgo-operator : Delete existing PGO Role Binding (Operator Namespace)] ****
changed: [localhost]

TASK [pgo-operator : Delete existing PGO Role (Operator Namespace)] ************
changed: [localhost]

TASK [pgo-operator : Ensure output directory exists] ***************************
ok: [localhost]

TASK [pgo-operator : Delete PGO client container] ******************************
changed: [localhost]

TASK [pgo-operator : Use kubectl or oc] ****************************************
ok: [localhost]

TASK [pgo-operator : include_tasks] ********************************************
included: /ansible/postgres-operator/roles/pgo-operator/tasks/namespace.yml for localhost

TASK [pgo-operator : Namespace List] *******************************************
ok: [localhost]

TASK [pgo-operator : Create Watched Namespaces] ********************************
skipping: [localhost] => (item=pgo) 
skipping: [localhost] => (item=db) 
skipping: [localhost] => (item=neo-staging) 
skipping: [localhost] => (item=neo-production) 

TASK [pgo-operator : Create Watched Namespaces (Reconcile RBAC)] ***************
skipping: [localhost] => (item=pgo) 
skipping: [localhost] => (item=db) 
skipping: [localhost] => (item=neo-staging) 
skipping: [localhost] => (item=neo-production) 

TASK [pgo-operator : Label Watched Namespaces (Reconcile RBAC)] ****************
skipping: [localhost] => (item=pgo) 
skipping: [localhost] => (item=db) 
skipping: [localhost] => (item=neo-staging) 
skipping: [localhost] => (item=neo-production) 

TASK [pgo-operator : Cleanup Local Namespace Target RBAC] **********************
skipping: [localhost] => (item=pgo) 
skipping: [localhost] => (item=db) 
skipping: [localhost] => (item=neo-staging) 
skipping: [localhost] => (item=neo-production) 

TASK [pgo-operator : Create Local Namespace Target RBAC] ***********************
skipping: [localhost] => (item=pgo) 
skipping: [localhost] => (item=db) 
skipping: [localhost] => (item=neo-staging) 
skipping: [localhost] => (item=neo-production) 

TASK [pgo-operator : Template Local Namespace RBAC] ****************************
skipping: [localhost]

TASK [pgo-operator : Cleanup Local Namespace Reconcile RBAC] *******************
skipping: [localhost] => (item=pgo) 
skipping: [localhost] => (item=db) 
skipping: [localhost] => (item=neo-staging) 
skipping: [localhost] => (item=neo-production) 

TASK [pgo-operator : Create Local Namespace Reconcile RBAC] ********************
skipping: [localhost] => (item=pgo) 
skipping: [localhost] => (item=db) 
skipping: [localhost] => (item=neo-staging) 
skipping: [localhost] => (item=neo-production) 

TASK [pgo-operator : Template PGO Admin Credentials] ***************************
changed: [localhost]

TASK [pgo-operator : Check PGO Admin Credentials] ******************************
changed: [localhost]

TASK [pgo-operator : Create PGO Admin Credentials] *****************************
changed: [localhost]

TASK [pgo-operator : Template PGO Admin Role & Permissions] ********************
changed: [localhost]

TASK [pgo-operator : Check PGO Admin Role & Permissions] ***********************
changed: [localhost]

TASK [pgo-operator : Create PGO Admin Role & Permissions] **********************
changed: [localhost]

TASK [pgo-operator : Template PGO Service Account] *****************************
changed: [localhost]

TASK [pgo-operator : Check PGO Service Account] ********************************
changed: [localhost]

TASK [pgo-operator : Create PGO Service Account] *******************************
changed: [localhost]

TASK [pgo-operator : Template Cluster RBAC (namespace_mode 'dynamic')] *********
changed: [localhost]

TASK [pgo-operator : Check Cluster RBAC (namespace_mode 'dynamic')] ************
changed: [localhost]

TASK [pgo-operator : Create Cluster RBAC (namespace_mode 'dynamic')] ***********
changed: [localhost]

TASK [pgo-operator : Template Cluster RBAC (namespace_mode 'readonly')] ********
skipping: [localhost]

TASK [pgo-operator : Check Cluster RBAC (namespace_mode 'readonly')] ***********
skipping: [localhost]

TASK [pgo-operator : Create Cluster RBAC (namespace_mode 'readonly')] **********
skipping: [localhost]

TASK [pgo-operator : Cluster Roles Disabled (namespace_mode 'disabled')] *******
skipping: [localhost]

TASK [pgo-operator : Create PGO Image Pull Secret] *****************************
skipping: [localhost]

TASK [pgo-operator : Check cluster-admin Cluster Role Binding for PGO Service Account] ***
skipping: [localhost]

TASK [pgo-operator : Create cluster-admin Cluster Role Binding for PGO Service Account] ***
skipping: [localhost]

TASK [pgo-operator : Template PGO RBAC] ****************************************
changed: [localhost]

TASK [pgo-operator : Check PGO RBAC] *******************************************
changed: [localhost]

TASK [pgo-operator : Create PGO RBAC] ******************************************
changed: [localhost]

TASK [pgo-operator : Template Local PGO User] **********************************
changed: [localhost]

TASK [pgo-operator : Check PGO BackRest Repo Secret] ***************************
changed: [localhost]

TASK [pgo-operator : Create PGO BackRest Repo Secret] **************************
changed: [localhost]

TASK [pgo-operator : Label PGO BackRest Repo Secret] ***************************
fatal: [localhost]: FAILED! => {"changed": true, "cmd": ["kubectl", "label", "secret", "generic", "-n", "pgo", "pgo-backrest-repo-config", "vendor=crunchydata"], "delta": "0:00:00.140298", "end": "2021-08-28 16:31:46.970387", "msg": "non-zero return code", "rc": 1, "start": "2021-08-28 16:31:46.830089", "stderr": "Error from server (NotFound): secrets \"generic\" not found", "stderr_lines": ["Error from server (NotFound): secrets \"generic\" not found"], "stdout": "secret/pgo-backrest-repo-config labeled", "stdout_lines": ["secret/pgo-backrest-repo-config labeled"]}

PLAY RECAP *********************************************************************
localhost                  : ok=55   changed=35   unreachable=0    failed=1    skipped=24   rescued=0    ignored=1   
```
</details>

**Comment by jkatz** [comment]
The [PGO upgrade process for v4](https://access.crunchydata.com/documentation/postgres-operator/latest/upgrade/) is defined here. Using Helm to upgrade is not fully supported. You would need to take some additional steps, likely following the manual guidance. I would recommend following the automatic procedure.

However, [v5](https://access.crunchydata.com/documentation/postgres-operator/v5/) will work with Helm upgrading.

**Comment by jkatz** [comment]
Also note that the tag is `centos8-13.4-4.7.2`. This was corrected in #2647 and may also be a cause of the issue.

**Comment by keithlayne** [comment]
Your docs say this:

https://access.crunchydata.com/documentation/postgres-operator/4.7.2/upgrade/automatedupgrade/#helm

Your helm chart has this:

https://github.com/CrunchyData/postgres-operator/blob/v4.7.2/installers/helm/values.yaml#L82

So, that tag came from postgres-operator, not me.

Your post-upgrade hook runs this:

```sh
kubectl label secret generic -n pgo pgo-backrest-repo-config vendor=crunchydata
```

...which is not legit.

Also, is there any upgrade path from 4 to 5?  I didn't see anything specific about that in the docs.

**Comment by jkatz** [comment]
@keithlayne I apologize, you are correct on the reference to the upgrade method. I have not spent as much time on the Helm parts as others on the project.

As I mentioned in my follow up, the principal issue is this line:

https://github.com/CrunchyData/postgres-operator/blob/v4.7.2/installers/helm/values.yaml#L43

This needs to be adjusted to `centos8-13.4-4.7.2`. That is where the `BackoffLimitExceeded` error comes from.

With the `kubectl label secret generic` you are correct that it is "not legit." I have pushed up a fix in #2650.

**Comment by keithlayne** [comment]
Sorry I pointed at the wrong tag in values.yaml there. Thanks for jumping on that.

**Comment by keithlayne** [comment]
Thanks for getting that merged @jkatz !  A big part of my mistake was using the 4.7.2 release instead of the `REL_4_7` branch. I think this could be easier on the user with something like #2231 implemented, but I understand helm is not a focus, at least in v4.

Not sure why GH didn't close this with your merge, but I'll go ahead and do so.

**Comment by jkatz** [comment]
I'll tag #2231 as something to consider for v5 given it'll likely be easier to implement + a better user experience overall with Helm. Thanks!

**Comment by keithlayne** [comment]
Sorry @jkatz , no dice. I just now had a chance to re-try this upgrade, and I get the same error.

Info that I hope is relevant:

- https://github.com/CrunchyData/postgres-operator/blob/REL_4_7/installers/helm/templates/_deployer_job_spec.yaml#L14 is where the post-upgrade job comes from. 
- It uses the `pgo-deployer` image at `.Values.pgo_image_tag`. 
- I don't override that, so I'm getting the default from `values.yaml`, which is `"centos8-4.7.2"` ([ref](https://github.com/CrunchyData/postgres-operator/blob/REL_4_7/installers/helm/values.yaml#L82))
- I didn't dig into your build tooling, and I thought it was odd that the change you made was in an ansible config. Maybe the deployer container leverages ansible? Anyway, I didn't see anywhere else in the repo that the secret labeling would have come from.
- The fact that that image tag has a `4.7.2` suffix implies to me either that it gets reused on changes in this branch automatically (probably not, considering the timing), I need to update the tag in my config,  or that image needs to be rebuilt and pushed to the container registry.
- The `BackoffLimitExceeded` error is coming from the post-upgrade job because the backoff is set to zero. The other image doesn't seem relevant to this error.

I hope that helps - please let me know if I can provide any more useful info.

**Comment by jkatz** [comment]
#2653 does resolve this issue. The fix will be available in 4.7.3, which we're planning to release in the not-too-distant future.
